### PR TITLE
make Fsa inherit from Array2 and add some operator overloads in Array2

### PR DESCRIPTION
**Pull request by qindazhu**

### CODE REVIEW

**Comment by csukuangfj** [comment]
+1

**Comment by qindazhu** [comment]
Fsa now inherits from Array2, I will replace Fsa with Array2 in those Fsa related algorithms one by one.

@danpovey @csukuangfj Please have a look.

**Comment by danpovey** [comment]
LGTM, thanks!

**Comment by danpovey** [comment]
OK with me

**Comment by csukuangfj** [comment]
+1

**Comment by qindazhu** [comment]
Thanks! Merging..

File: k2/csrc/fsa.h [hunk 1]
```diff
@@ -79,9 +79,18 @@ struct ArcHash {
...
+  TODO(haowen): add below comments in the final version:
+    (In Array2 representation) We represent an empty FSA(the one that accepts no
+  strings) by having no states at all, so `size1` would be 0 (As an empty FSA is
+  an initialized Array2 object, so `indexes` would be allocated and has at least
+  one element, but we don't care about it here).
  */
-struct Fsa {
+// TODO(haowen): finally we will remove `arc_indexes` and `arcs`, but for now,
+// we would keep them to replace Fsa with Array2 incrementally.
+struct Fsa : public Array2<Arc *, int32_t> {
```
Comment (csukuangfj) [inline k2/csrc/fsa.h:93]: Does this mean that the stride of the underlying memory is always 1 ?

Comment (qindazhu) [reply k2/csrc/fsa.h:93]: yes, for `FSA`.

**Pull request merged**